### ISSUE TITLE
fairseq.data.iterators | Data loading buffer is empty or nearly empty. This may indicate a data loading bottleneck, and increasing the number of workers (--num-workers) may help.

### ISSUE THREAD

**Issue by fenimi** [pytorch/fairseq#2264]
### Before asking:   
1. search the issues.   
2. search the docs.    

#### What is your question?
How can I solve this problem? This warning shows up after a few iterations and it stops training:

fairseq.data.iterators | Data loading buffer is empty or nearly empty. This may indicate a data loading bottleneck, and increasing the number of workers (--num-workers) may help. 

#### Code

   
#!/bin/bash
#SBATCH --account=XXXXXX
#SBATCH --gres=gpu:v100l:1       # Number of GPUs (per node)
#SBATCH --mem=100G               # memory
#SBATCH --time=06-59:59           # time (DD-HH:MM)           
#SBATCH --job-name=script_ca-es
#SBATCH --output=script_ca-es.out

CUDA_VISIBLE_DEVICES=0 fairseq-train data-bin_ca-es/tokenized_ca-es/ --arch transformer --save-dir checkpoints_ca-es  --share-decoder-input-output-embed     --optimizer adam --adam-betas '(0.9,0.98)' --clip-norm 0.0 --lr 5e-4 --lr-scheduler inverse_sqrt --warmup-updates 4000 --dropout 0.3 --weight-decay 0.0001 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 --max-tokens 4096 --eval-bleu --eval-bleu-args '{"beam": 5, "max_len_a": 1.2, "max_len_b": 10}' --eval-bleu-detok moses --eval-bleu-print-samples --best-checkpoint-metric bleu --maximize-best-checkpoint-metric
#### What have you tried?
I've changed the --num-workers when preparing the preprocessing step. 

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master):
 - PyTorch Version (e.g., 1.0): 1.4
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source): git clone
 - Build command you used (if compiling from source): 
 - Python version: 3.6.3
 - CUDA/cuDNN version: 8.0.44
 - GPU models and configuration:
 - Any other relevant information:

**Comment by myleott** [comment]
Hmm, it was previously printing these too frequently, but that should have been fixed in 14ee059. Can you confirm you're using a commit more recent than that?

**Comment by fenimi** [comment]
> Hmm, it was previously printing these too frequently, but that should have been fixed in [14ee059](https://github.com/pytorch/fairseq/commit/14ee059a36092a4216ec38601ca32be11383eecb). Can you confirm you're using a commit more recent than that?

I'm using an older version I believe since I downloaded this about 3 weeks ago. Do I just reclone the github to have the updates on my current system?

**Comment by myleott** [comment]
Yep recloning works

**Comment by hljjjmssyh** [comment]
@fenimi Hi! Have you solved this problem？

**Comment by fenimi** [comment]
Hi! I'm still running, I haven't encountered it yet. Like I said it runs for a while before giving errors. I'll post an update when the training is complete

**Comment by fenimi** [comment]
Hi, so regarding the errors, I no longer get the error. However, after loading the last checkpoint, it stops printing anything which I find strange.  It prints this out:

_**2020-07-05 19:40:49 | INFO | fairseq.trainer | loading train data for epoch 9
2020-07-05 19:40:57 | INFO | fairseq.data.data_utils | loaded 49747362 examples from: data-bin_multi-left-right/multilingual-left-right/train.left-right.left
2020-07-05 19:41:05 | INFO | fairseq.data.data_utils | loaded 49747362 examples from: data-bin_multi-left-right/multilingual-left-right/train.left-right.right
2020-07-05 19:41:05 | INFO | fairseq.tasks.translation | data-bin_multi-left-right/multilingual-left-right train left-right 49747362 examples**_

I'm not sure why though.
It is still running the model and in about 24 hours nothing has been printed.

**Comment by hljjjmssyh** [comment]
> Hi, so regarding the errors, I no longer get the error. However, after loading the last checkpoint, it stops printing anything which I find strange. It prints this out:
> 
> _**2020-07-05 19:40:49 | INFO | fairseq.trainer | loading train data for epoch 9 2020-07-05 19:40:57 | INFO | fairseq.data.data_utils | loaded 49747362 examples from: data-bin_multi-left-right/multilingual-left-right/train.left-right.left 2020-07-05 19:41:05 | INFO | fairseq.data.data_utils | loaded 49747362 examples from: data-bin_multi-left-right/multilingual-left-right/train.left-right.right 2020-07-05 19:41:05 | INFO | fairseq.tasks.translation | data-bin_multi-left-right/multilingual-left-right train left-right 49747362 examples**_
> 
> I'm not sure why though.
> It is still running the model and in about 24 hours nothing has been printed.

I solved this problem by using an old version（v0.9.0), and it works well, although this is not the best solution.

**Comment by fenimi** [comment]
> > Hi, so regarding the errors, I no longer get the error. However, after loading the last checkpoint, it stops printing anything which I find strange. It prints this out:
> > _**2020-07-05 19:40:49 | INFO | fairseq.trainer | loading train data for epoch 9 2020-07-05 19:40:57 | INFO | fairseq.data.data_utils | loaded 49747362 examples from: data-bin_multi-left-right/multilingual-left-right/train.left-right.left 2020-07-05 19:41:05 | INFO | fairseq.data.data_utils | loaded 49747362 examples from: data-bin_multi-left-right/multilingual-left-right/train.left-right.right 2020-07-05 19:41:05 | INFO | fairseq.tasks.translation | data-bin_multi-left-right/multilingual-left-right train left-right 49747362 examples**_
> > I'm not sure why though.
> > It is still running the model and in about 24 hours nothing has been printed.
> 
> I solved this problem by using an old version（v0.9.0), and it works well, although this is not the best solution.

oh okay, thank you for that information.

**Comment by atreyasha** [comment]
I installed `fairseq` from commit 9f92b05 (master `HEAD` as of now) and am still encountering the same data iterator warnings as the OP. This warning still persists even after increasing `--num-workers` from 5 up to 20. The script still runs fine but every ~500 updates or so it returns the same warning. Could this be solely attributed to data loading issues or perhaps something else?

Following is the base command I was executing, you can just ignore the shell variables as they are just used for naming stuff:

```shell
fairseq-train \
      "data/wmt16_en_de_bpe32k/bin" \
      --arch "$arch" --share-all-embeddings \
      --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
      --lr 0.001 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 \
      --dropout 0.3 --weight-decay 0.0 \
      --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
      --max-tokens 3584 --fp16 --update-freq 16 \
      --keep-last-epochs 10 --patience 3 --num-workers 5 \
      --save-dir "./models/${arch}.wmt16.de-en.${unix_epoch}" \
      --tensorboard-logdir "./models/${arch}.wmt16.de-en.${unix_epoch}"
```

PS: I can also verify that I was not receiving this warning in the `v0.9.0` release.
